core/src/commonMain/kotlin/com/xebia/functional/xef/auto/AIScope.kt

```diff
@@ -0,0 +1,501 @@
+import com.xebia.functional.tokenizer.ModelType
+import com.xebia.functional.tokenizer.truncateText
+import com.xebia.functional.xef.AIError
+import com.xebia.functional.xef.embeddings.Embeddings
+import com.xebia.functional.xef.llm.openai.*
+import com.xebia.functional.xef.llm.openai.functions.CFunction
+import com.xebia.functional.xef.llm.openai.images.ImagesGenerationRequest
+import com.xebia.functional.xef.llm.openai.images.ImagesGenerationResponse
+import com.xebia.functional.xef.prompt.Prompt
+import com.xebia.functional.xef.vectorstores.CombinedVectorStore
+import com.xebia.functional.xef.vectorstores.LocalVectorStore
+import com.xebia.functional.xef.vectorstores.VectorStore
+import io.github.oshai.kotlinlogging.KLogger
+import io.github.oshai.kotlinlogging.KotlinLogging
+import kotlin.jvm.JvmName
+
+/**
+ * The [AIScope] is the context in which [AI] values are run. It encapsulates all the dependencies
+ * required to run [AI] values, and provides convenient syntax for writing [AI] based programs.
+ */
+class AIScope(
+  val openAIClient: OpenAIClient,
+  val context: VectorStore,
+  val embeddings: Embeddings,
+  val logger: KLogger = KotlinLogging.logger {}
+) {
+
+  /**
+   * Allows invoking [AI] values in the context of this [AIScope].
+   *
+   * ```kotlin
+   * data class CovidNews(val title: String, val content: String)
+   * val covidNewsToday = ai {
+   *   val now = LocalDateTime.now()
+   *   agent(search("$now covid-19 News")) {
+   *     prompt<CovidNews>("write a paragraph of about 300 words about the latest news on covid-19 on $now")
+   *   }
+   * }
+   *
+   * data class BreakingNews(val title: String, val content: String, val date: String)
+   *
+   * fun breakingNews(date: LocalDateTime): AI<BreakingNews> = ai {
+   *   agent(search("$date Breaking News")) {
+   *     prompt("Summarize all breaking news that happened on ${now.minusDays(it)} in about 300 words")
+   *   }
+   * }
+   *
+   * suspend fun AIScope.breakingNewsLastWeek(): List<BreakingNews> {
+   *   val now = LocalDateTime.now()
+   *   return (0..7).parMap { breakingNews(now.minusDays(it)).invoke() }
+   * }
+   *
+   * fun news(): AI<List<News>> = ai {
+   *   val covidNews = parZip(
+   *     { covidNewsToday() },
+   *     { breakingNewsLastWeek() }
+   *   ) { covidNews, breakingNews -> listOf(covidNews) + breakingNews }
+   * }
+   * ```
+   */
+  @AiDsl @JvmName("invokeAI") suspend operator fun <A> AI<A>.invoke(): A = invoke(this@AIScope)
+
+  @AiDsl
+  suspend fun extendContext(vararg docs: String) {
+    context.addTexts(docs.toList())
+  }
+
+  /**
+   * Creates a nested scope that combines the provided [store] with the outer _store_. This is done
+   * using [CombinedVectorStore].
+   *
+   * **Note:** if the implementation of [VectorStore] is relying on resources you're manually
+   * responsible for closing any potential resources.
+   */
+  @AiDsl
+  suspend fun <A> contextScope(store: VectorStore, block: AI<A>): A =
```

Comment:
This probably another candidate for the `AsFuture` plugin 🤔 Where the `block` argument needs to be translated from `suspend AIScope.() -> A` to `AIScope.() -> Future<A>`.

`AIScope.() -> Future<A>` will appear as `(AIScope) -> Future<A>` from Java.